feat[next]: DaCe scan operator

edopao opened the pull request:
## Description

Add support to DaCe backend for ITIR scan operator.

## Requirements

- [ ] All fixes and/or new features come with corresponding tests.
- [ ] Important design decisions have been documented in the approriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/Index.md) folder.

If this PR contains code authored by new contributors please make sure:

- [x] All the authors are covered by a valid contributor assignment agreement provided to ETH Zurich and signed by the employer if needed.
- [x] The PR contains an updated version of the `AUTHORS.md` file adding the names of all the new contributors.

petiaccja commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py:
```diff
@@ -596,3 +600,33 @@ def closure_to_tasklet_sdfg(
     return context, outputs
+
+
+def lambda_to_tasklet_sdfg(
```
We should ideally unify `lambda_to_tasklet_sdfg` and `closure_to_tasklet_sdfg`, they more or less serve the same purpose. Not sure what's the best approach.

edopao replied:
Yes, they serve the same purpose, but the representation of the lambda function is different.

petiaccja replied:
Addressed in your last commit / discussion below. Closing.

petiaccja commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py:
```diff
@@ -596,3 +600,33 @@ def closure_to_tasklet_sdfg(
+        stencil_fobj = cast(FunCall, node.stencil)
+        return cast(Lambda, stencil_fobj.args[0])
+
+    call = itir.Lambda(expr=get_lambda().expr, params=get_lambda().params)
```
```suggestion
    callable = itir.Lambda(expr=get_lambda().expr, params=get_lambda().params)
```

edopao replied:
`A001 variable "callable" is shadowing a python builtin`

Instead, I can use `funobj`, for function object.

petiaccja commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py:
```diff
@@ -596,3 +600,33 @@ def closure_to_tasklet_sdfg(
+
+    def get_lambda() -> Lambda:
+        stencil_fobj = cast(FunCall, node.stencil)
+        return cast(Lambda, stencil_fobj.args[0])
```
In theory, the callee could be a `Lambda` as well as a `SymRef` to a `FunctionDefinition`, so it's dangerous to cast it to a `Lambda` without checking. While as of of now probably all will be just lambdas, we should handle the different callables uniformly.

edopao replied:
However, this cast happens in a function for visiting lambda functions -- `lambda_to_tasklet_sdfg`. Is it OK if I just add an assert on the instance type?
```suggestion
        assert isinstance(stencil_fobj.args[0], Lambda)
        return stencil_fobj.args[0]
```

petiaccja replied:
I think I'd probably split this function (`lambda_to_tasklet_sdfg`) into two:
- A function that really just turns a lambda into a tasklet SDFG -- this one you could reuse with the parallel case
- A function that extracts the lambda function from the scan `StencilClosure`
(- Another function that extracts the lambda function/func def from the parallel `StencilClosure`)

The problem is that it does not simply convert a lambda to a tasklet SDFG, but rather it extracts the lambda from a scan stencil closure and then it converts it to an SDFG, which breaks the single responsibility principle and make the code not super reusable.

edopao replied:
@petiaccja I have tried to address your comment in the latest commit

petiaccja replied:
Yeah, that looks better to me. What do you think?

edopao replied:
I agree on the single responsibility principle, and I like that we are now using one common function `closure_to_tasklet_sdfg`. It would be even nicer if the visitor functions would have the same return signature, but unfortunately this is not the case: the node `visit` function returns `context, inner_inputs, inner_outputs` in case of lambda function, only `inner_outputs` in case of function call.

petiaccja commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -281,6 +330,212 @@ def visit_StencilClosure(
+        closure_domain: tuple[tuple[str, tuple[ValueExpr, ValueExpr]], ...],
+    ) -> tuple[dace.SDFG, dict[str, str | dace.subsets.Subset], list[str], int]:
+        # extract scan arguments
+        def get_scan_args() -> tuple[Literal, Literal]:
+            stencil_fobj = cast(FunCall, node.stencil)
+            return cast(Literal, stencil_fobj.args[1]), cast(Literal, stencil_fobj.args[2])
+
+        # select the scan dimension based on program argument for column axis
+        def get_scan_dim() -> tuple[str, int, ScalarType]:
+            assert isinstance(node.output, SymRef)
+            field_type = cast(FieldType, self.storages[node.output.id])
+            assert isinstance(self.column_axis, Dimension)
+            return (
+                self.column_axis.value,
+                field_type.dims.index(self.column_axis),
+                field_type.dtype,
+            )
```
I'd extract these to the enclosing module as private functions, or if not possible, find other ways to break up the enclosing function. These utility functions are often good candidates for pure functions that have no side effects, thus always return the same results for the same inputs. Pure function are the best for testing, but by making them nested function and relying on capturing as opposed to function parameters for inputs you eliminate this ease of testing.

edopao replied:
Good point about the testability aspect, I will try to make pure functions in a nice way.

tehrengruber commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py:
```diff
@@ -406,7 +407,10 @@ def _visit_call(self, node: itir.FunCall):
 
     def _visit_deref(self, node: itir.FunCall) -> list[ValueExpr]:
-        iterator: IteratorExpr = self.visit(node.args[0])
+        iterator = self.visit(node.args[0])
+        if not isinstance(iterator, IteratorExpr):
+            # already a list of ValueExpr
+            return iterator
```
Yes, deref only accepts an iterator.

edopao replied:
The reason why I am getting `ValueExpr` is that I manually create some IR nodes for the input arguments to the lambda function. I will add a comment to point to the source location.

gronerl replied:
So this hack should actually lead to an invalid graph. Consider this graph:
![image](https://github.com/GridTools/gt4py/assets/22642725/8a674199-5322-4f48-8da3-407e9504283e)

Here, scalars are created (`__var_45`, `__var_57`) but they are only used in a memlet subset, where only symbols would be allowed. This is a bug in dace validation and that it works in code generation is coincidental.

Instead, those scalars and their accesses, edges etc should not be created and instead, the value they are initialized with in the tasklet (`__out1_size_2`, `0`) should be used in the subset directly.

edopao replied:
In the sdfg for `test_scalar_scan`, the lambda function has three arguments:
```
params = {list: 3} [Sym(id=SymbolName('state'), kind=None, dtype=None), Sym(id=SymbolName('qc_in'), kind=None, dtype=None), Sym(id=SymbolName('scalar'), kind=None, dtype=None)]
 0 = {Sym} Sym(id=SymbolName('state'), kind=None, dtype=None)
 1 = {Sym} Sym(id=SymbolName('qc_in'), kind=None, dtype=None)
 2 = {Sym} Sym(id=SymbolName('scalar'), kind=None, dtype=None)
```
 Your comment applies to `scalar`, which you are saying should be handled as a symbol (the way it looks now depends on the way it is implemented in baseline). But what about `qc_in` and `state`?

edopao replied:
This bug was fixed in [494db45](https://github.com/GridTools/gt4py/pull/1284/commits/494db45ca2a81b8bc58dfe86026508026110cc20)

edopao commented on ci/base.Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM docker.io/nvidia/cuda:11.2.0-devel-ubuntu20.04
+FROM docker.io/nvidia/cuda:11.2.2-devel-ubuntu20.04
```
Right, this change was merged yesterday into main as a separate PR.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/__init__.py:
```diff
@@ -71,20 +71,27 @@ def get_stride_args(
 
 @program_executor
 def run_dace_iterator(program: itir.FencilDefinition, *args, **kwargs) -> None:
+    column_axis = None
+    if "column_axis" in kwargs:
+        column_axis = kwargs["column_axis"]
```
Thanks, nice API.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/__init__.py:
```diff
@@ -71,20 +71,27 @@ def get_stride_args(
     arrays: Mapping[str, dace.data.Array], args: Mapping[str, Any]
 ) -> dict[str, Any]:
     return {
-        str(sym): size / value.itemsize
+        str(sym): size // value.itemsize
         for name, value in args.items()
         for sym, size in zip(arrays[name].strides, value.strides)
```
Good idea, I'll fix it.

gronerl replied:
Unfortunately, this is not the case:
```
>>> type(int(4)/int(4))
<class 'float'>
```
And this actually causes warnings to be printed when not done with floordiv. Renaming to stride seems right though.

edopao replied:
Correct. In my change, I have used `divmod`
```
q, r = divmod(int(4), int(4))
type(q)
<class 'int'>
```

petiaccja replied:
Sounds good!

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -91,7 +97,13 @@ def visit_FencilDefinition(self, node: itir.FencilDefinition):
             assert isinstance(closure.output, itir.SymRef)
 
-            input_names = [str(inp.id) for inp in closure.inputs]
+            # arguments with scalar type are passed as symbols
+            input_names = list(
+                filter(
+                    lambda name: isinstance(self.storages[name], ts.FieldType),
+                    [str(inp.id) for inp in closure.inputs],
+                )
+            )
```
I agree, it looks nicer.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py:
```diff
@@ -596,3 +600,33 @@ def closure_to_tasklet_sdfg(
+    # add lambda parameters to symbol map
+    for p in get_lambda().params:
+        pnode = state.add_access(p.id)
+        symbol_map[p.id] = ValueExpr(value=pnode.data, dtype=dace.float64)
```
@petiaccja @tehrengruber Here is where I create the `ValueExpr` nodes. I agree, this could be a problem. In particular, see the hard-coding of float64 type, which obviously is a quick and dirty solution. Can you suggest a better way?

edopao replied:
It is possible to retrieve the data type for the lambda input arguments:
```suggestion
        pnode_type = node_types[id(p)]
        symbol_map[p.id] = ValueExpr(value=pnode.data, dtype=itir_type_as_dace_type(pnode_type.dtype))
```

edopao replied:
We have found a better solution, this code was removed.

edopao commented:
@petiaccja Thank you for the review. I have addressed most comments and resolved the corresponding conversations. I have just left a few conversations open for follow-up discussion.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/__init__.py:
```diff
@@ -70,21 +70,32 @@ def get_shape_args(
-        for name, value in args.items()
-        for sym, size in zip(arrays[name].strides, value.strides)
-    }
+    stride_args = {}
```
I guess that the type annotation can be derived directly from the return type of the function, in this case. Adding the type annotation on this variable seems a bit redundant to me. One thing I react about, now when I look more carefully at type annotations in this function, is that the type of the return value should be a bit more specific to make it useful. I am going to change it to `dict[str, int]` (and same for shape) since shape and stride are integer values.

edopao replied:
I guess that the type annotation can be derived directly from the return type of the function, in this case. Adding the type annotation on this variable seems a bit redundant to me.
One thing I react about, now when I look more carefully at type annotations in this function, is that the type of the return value should be a bit more specific to make it useful. I am going to change it to dict[str, int] (and same for shape) since shape and stride are integer values.

petiaccja replied:
Yeah, that's what I was guessing as well. In that case it's good like this.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/__init__.py:
```diff
@@ -70,21 +70,32 @@ def get_shape_args(
+    for name, value in args.items():
+        for sym, stride_size in zip(arrays[name].strides, value.strides):
+            stride, remainder = divmod(stride_size, value.itemsize)
+            if remainder:
```
I agree, I'll fix this.

edopao replied:
I agree, I'll fix this.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -40,19 +42,40 @@
 
+def get_scan_args(stencil: Expr) -> tuple[Literal, Literal]:
+    stencil_fobj = cast(FunCall, stencil)
+    return cast(Literal, stencil_fobj.args[1]), cast(Literal, stencil_fobj.args[2])
```
Accepted. Let's also check that the first argument is a bool in this case.

edopao replied:
Accepted. Let's also check that the first argument is a bool in this case.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -40,19 +42,40 @@
+    return cast(Literal, stencil_fobj.args[1]), cast(Literal, stencil_fobj.args[2])
+
+
+def get_scan_dim(
```
Ok

edopao replied:
Ok

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -40,19 +42,40 @@
     def __init__(
         self,
         param_types: list[ts.TypeSpec],
+        column_axis: Optional[Dimension],
```
Right, it makes sense.

edopao replied:
Right, it makes sense.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -172,114 +204,346 @@ def visit_StencilClosure(
         input_memlets = [
-            create_memlet_full(name, closure_sdfg.arrays[name]) for name in input_names
+            create_memlet_full(name, closure_sdfg.arrays[name])
+            for name in map(
+                lambda x: x
+                if isinstance(self.storages[x], ts.FieldType)
+                else cast(ValueExpr, program_arg_syms[x]).value.data,
+                input_names,
+            )
```
I agree, done.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -172,114 +204,346 @@ def visit_StencilClosure(
+        self,
+        node: itir.StencilClosure,
+        array_table: dict[str, dace.data.Array],
+        closure_domain: tuple[tuple[str, tuple[ValueExpr, ValueExpr]], ...],
```
Good catch. Shouldn't type-checking have caught it?

petiaccja replied:
I don't think so, because you are not doing anything illegal typing-wise. Maybe it shows a warning, but I think mypy only shows warnings when there is an error as well.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -172,114 +204,346 @@ def visit_StencilClosure(
+
+        if scan_forward_arg.value == "True":  # forward scan
+            scan_sdfg.add_loop(
+                start_state,
+                lambda_state,
+                end_state,
+                loop_var=f"i_{scan_dim}",
+                initialize_expr=f"{scan_lb_str}",
+                condition_expr=f"i_{scan_dim} < {scan_ub_str}",
+                increment_expr=f"i_{scan_dim} + 1",
             )
-            closure_state.add_edge(
-                ub.value, None, map_entry, ub.value.data, create_memlet_at(ub.value.data, ("0",))
+        else:
+            assert scan_forward_arg.value == "False"
+            scan_sdfg.add_loop(
+                start_state,
+                lambda_state,
+                end_state,
+                loop_var=f"i_{scan_dim}",
+                initialize_expr=f"{scan_ub_str} - 1",
+                condition_expr=f"i_{scan_dim} >= {scan_lb_str}",
+                increment_expr=f"i_{scan_dim} - 1",
             )
```
Good idea!

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/__init__.py:
```diff
@@ -70,21 +70,32 @@ def get_shape_args(
+            stride, remainder = divmod(stride_size, value.itemsize)
+            if remainder:
+                raise ValueError(
+                    f"Invalid stride for argument {sym}: {stride_size} / {value.itemsize}"
```
Right, I'll fix this.

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -40,19 +42,40 @@
+
+def get_scan_dim(
+    column_axis: Dimension,
+    storages: dict[str, ts.TypeSpec],
```
Done

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -40,19 +42,40 @@
+
 class ItirToSDFG(eve.NodeVisitor):
     param_types: list[ts.TypeSpec]
     storages: dict[str, ts.TypeSpec]
```
Done

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -172,114 +204,346 @@ def visit_StencilClosure(
+        # connect access nodes to lambda inputs
+        lambda_input_subset = ", ".join([f"i_{dim}" for dim, _ in closure_domain])
+        for (inner_name, _), data_name in zip(lambda_inputs[1:], input_names):
+            if isinstance(self.storages[data_name], ts.FieldType):
+                lambda_state.add_memlet_path(
+                    lambda_state.add_access(data_name),
+                    scan_inner_node,
+                    memlet=dace.Memlet(data=f"{data_name}", subset=lambda_input_subset),
+                    src_conn=None,
+                    dst_conn=inner_name,
+                )
+            else:
+                lambda_state.add_memlet_path(
+                    lambda_state.add_access(data_name),
+                    scan_inner_node,
+                    memlet=dace.Memlet(data=f"{data_name}", subset="0"),
+                    src_conn=None,
+                    dst_conn=inner_name,
```
Done

edopao commented on src/gt4py/next/program_processors/runners/dace_iterator/itir_to_sdfg.py:
```diff
@@ -172,114 +204,346 @@ def visit_StencilClosure(
+        self,
+        node: itir.StencilClosure,
+        array_table: dict[str, dace.data.Array],
+        closure_domain: tuple[tuple[str, tuple[ValueExpr, ValueExpr]], ...],
```
Done

edopao reviewed (commented):
Half-way in the process of answering your comments, I have learnt to answer all in a single review :)
